libnopegl/src/text.c

```diff
@@ -308,14 +308,53 @@ static int set_vec3_value(float *dst, struct ngl_node *node, const float *value,
     return 0;
 }
 
-static int set_transform(float *dst, struct ngl_node *node, double t)
+static int set_transform(float *dst, struct texteffect_opts *effect_opts,
+                         struct ngli_box box, struct ngli_box chr_box,
+                         const struct char_info *chr, double t)
 {
+    struct ngl_node *node = effect_opts->transform_chain;
+
     if (!node)
         return 0;
     int ret = ngli_node_update(node, t);
     if (ret < 0)
         return ret;
-    ngli_transform_chain_compute(node, dst);
+
+    float anchor_x, anchor_y;
+    if (effect_opts->anchor_ref == NGLI_TEXT_ANCHOR_REF_CHAR) {
+        /*
+         * Go to the the center of the character quad, then adjust to honor the
+         * anchor parameter according to the real dimension of the character
+         */
+        anchor_x = chr_box.x + chr_box.w / 2.f + effect_opts->anchor[0] * chr->real_dim[0] / 2.f;
+        anchor_y = chr_box.y + chr_box.h / 2.f + effect_opts->anchor[1] * chr->real_dim[1] / 2.f;
+    } else if (effect_opts->anchor_ref == NGLI_TEXT_ANCHOR_REF_BOX) {
+        /* Remap an anchor in [-1,1] to the text bounding box coordinates */
+        const float norm_x = NGLI_LINEAR_NORM(-1.f, 1.f, effect_opts->anchor[0]);
+        const float norm_y = NGLI_LINEAR_NORM(-1.f, 1.f, effect_opts->anchor[1]);
+        anchor_x = NGLI_MIX_F32(box.x, box.x + box.w, norm_x);
+        anchor_y = NGLI_MIX_F32(box.y, box.y + box.h, norm_y);
+    } else if (effect_opts->anchor_ref == NGLI_TEXT_ANCHOR_REF_VIEWPORT) {
+        anchor_x = effect_opts->anchor[0];
+        anchor_y = effect_opts->anchor[1];
+    } else {
+        ngli_assert(0);
+    }
+
+    NGLI_ALIGNED_MAT(tm);
+    NGLI_ALIGNED_MAT(treloc0);
+    NGLI_ALIGNED_MAT(treloc1);
```

Comment:
nit: make the names consistent (if that makes sense). I guess `tm` stands for `transformation matrix` while `treloc0` stands for transformation effect location 0 ?

Reply:
`tm` for "transform matrix" yes, `treloc` for "transform relocation"

What alternative do you suggest? `tmreloc*`?

Reply:
`tmreloc*` sounds good.